src/main/java/com/iota/iri/network/pipeline/SolidifyStage.java

```diff
@@ -0,0 +1,87 @@
+ * transactions get requested. If the transaction is unsolid, a random solid tip is broadcast instead to keep the
+ * requests transmitting to neighbors.
+ */
+public class SolidifyStage implements Stage {
+    private static final Logger log = LoggerFactory.getLogger(SolidifyStage.class);
+
+    private TransactionSolidifier txSolidifier;
+    private TransactionValidator txValidator;
+    private TipsViewModel tipsViewModel;
+    private Tangle tangle;
+
+    /**
+     * Constructor for the {@link SolidifyStage}.
+     *
+     * @param txSolidifier      Transaction solidifier implementation for updating transaction's solidity status
+     * @param txValidator       Transaction validator implementation for determining the validity of a transaction
+     * @param tipsViewModel     Used for broadcasting random solid tips if the subject transaction is unsolid
+     * @param tangle            A reference to the nodes DB
+     */
+    public SolidifyStage(TransactionSolidifier txSolidifier, TransactionValidator txValidator,
+                         TipsViewModel tipsViewModel, Tangle tangle){
+        this.txSolidifier = txSolidifier;
+        this.txValidator = txValidator;
+        this.tipsViewModel = tipsViewModel;
+        this.tangle = tangle;
+    }
+
+    /**
+     * Processes the payload of the {@link ProcessingContext} as a {@link SolidifyPayload}. First the transaction will
+     * be checked for solidity and validity. If the transaction is already solid or can be set solid quickly by the
+     * transaction validator, the transaction is passed to the {@link BroadcastStage}. If not, the transaction is
+     * added to the solidification queue, and a random solid tip is pulled form the {@link TipsViewModel} to be
+     * broadcast instead.
+     *
+     * @param ctx       The context to process
+     * @return          The output context, in most cases a {@link BroadcastPayload}.
+     */
+    @Override
+    public ProcessingContext process(ProcessingContext ctx){
+        try {
+            SolidifyPayload payload = (SolidifyPayload) ctx.getPayload();
+            TransactionViewModel tvm = payload.getTransaction();
+
+            if (tvm.isSolid() || txValidator.quickSetSolid(tvm)) {
+                ctx.setNextStage(TransactionProcessingPipeline.Stage.BROADCAST);
+                ctx.setPayload(new BroadcastPayload(payload.getOriginNeighbor(), payload.getTransaction()));
+                return ctx;
+            }
+
+            txSolidifier.addToSolidificationQueue(tvm.getHash());
+
+            Hash tipHash = tipsViewModel.getRandomSolidTipHash();
+            if(tipHash != null) {
+                TransactionViewModel solidTip = fromHash(tangle, tipHash);
+                ctx.setNextStage(TransactionProcessingPipeline.Stage.BROADCAST);
+                ctx.setPayload(new BroadcastPayload(payload.getOriginNeighbor(), solidTip));
+                return ctx;
+            }
```

Comment:
aren't we in danger of broadcasting a tip more than once?

Reply:
We are, but since transaction requesting is tied directly to broadcasting, it lessens performance and sync speed to avoid sending these extra broadcasts. At least in this instance it will be mostly solid transactions being broadcast. I'm going to add a solidity check to the `solidTip` though to make sure that it isn't broadcasting any unsolid tips returned by the `tipsViewModel`.

Reply:
If we refactor transaction requesting to be delineated from the broadcast stage, then we wouldn't need little things like this to keep the transactions flowing.